From a768924cb6aa8aed4e25acf802eaedb064667e44 Mon Sep 17 00:00:00 2001
From: sai surya teja <sssteja@habana.ai>
Date: Wed, 12 Apr 2023 12:28:25 +0300
Subject: [PATCH] [SW-140527] gaudi2: security: reset unused lbw rrs in erom

This patch resets unused LBW RRs in EROM to avoid any compatibility
related issues between different firmware versions.

Change-Id: Ife41875b8a8759d465f5f25ebdc80d21cbee6c15
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 319b284..7a87203 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -155,6 +155,24 @@
 	.rr_en = mmDCORE0_HMMU0_MMU_DDR_RANGE_REG_ENABLE,
 };
 
+static const uint8_t hl_sec_unused_common_lbw_rr_list[] = {
+	HL_SEC_LBW_PCIE_MSI,
+	HL_SEC_LBW_CPU_IF,  /* unused in EROM. used in preboot. Hence keeping it here */
+	HL_SEC_LBW_PSOC_MASTER_IF,
+	HL_SEC_LBW_PCIE_PMA_01,
+	HL_SEC_LBW_PCIE_PMA_23,
+	HL_SEC_LBW_ELBI_FIT_REGION, /* unused in EROM. used in preboot. Hence Keeping it here */
+	HL_SEC_LBW_CPU_CA53_CFG,
+	HL_SEC_LBW_CPU_TIMESTAMP_MSTR_IF,
+	HL_SEC_LBW_PSOC_ARC_DBG,
+	HL_SEC_LBW_PSOC_CA53_DBG,
+	HL_SEC_LBW_PSOC_PID,
+	HL_SEC_LBW_DBI_1,
+	HL_SEC_LBW_DBI_2,
+	HL_SEC_LBW_PSOC_ETR_RWD,
+	HL_SEC_LBW_RSVD
+};
+
 #define SRAM_RSVD_BASE		0x1000008000000000ull
 #define SRAM_END		(SRAM_BASE_ADDR + SRAM_SIZE)
 #define SRAM_END_EXC_TOP4K	(SRAM_END - 4096) /* Short HBW RR can't protect last 4K of SRAM */
@@ -866,11 +884,28 @@
 			     HL_SEC_PCIE_INST_NUM, &hl_sec_lbw_pcie_rrs);
 }
 
+static void preboot_reset_unused_lbw_rrs(void)
+{
+	struct hl_rr_eng_select rr_sel;
+	uint8_t i;
+
+	memset(&rr_sel, 0, sizeof(struct hl_rr_eng_select));
+
+	rr_sel.pcie = 1;
+	rr_sel.noc = 1;
+	rr_sel.psoc = 0; /* HL_SEC_HBW_PSOC_LBW is used for PSOC by default. No need to clear */
+	for (i = 0; i < ARRAY_SIZE(hl_sec_unused_common_lbw_rr_list); i++)
+		hl_secure_lbw_rr(hl_sec_unused_common_lbw_rr_list[i], 0, 0, rr_sel);
+}
+
 static void preboot_secure_lbw(void)
 {
 	u64 min, max;
 	struct hl_rr_eng_select rr_sel;
 
+	/* NOTE: make sure unused RR list is modified whenever there is RR config change below */
+	preboot_reset_unused_lbw_rrs();
+
 	memset(&rr_sel, 0, sizeof(struct hl_rr_eng_select));
 
 	/*
